File: cas/src/main/scala/smile/cas/Vector.scala
/*******************************************************************************
 * Copyright (c) 2010-2019 Haifeng Li
 *
 * Smile is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Smile is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Smile.  If not, see <https://www.gnu.org/licenses/>.
 *******************************************************************************/

package smile.cas

/** Vector: rank-1 tensor. */
trait Vector extends Tensor {
  override def rank: Option[Int] = Some(1)
  override def shape: Option[Array[IntScalar]] = Some(Array(size))

  /** The size of vector. */
  def size: IntScalar

  /** Applies the expression. */
  def apply(env: Map[String, Tensor]): Vector
  /** Applies the expression. */
  def apply(env: (String, Tensor)*): Vector = apply(Map(env: _*))
  /** Simplify the expression. */
  def simplify: Vector = this

  /** Returns the partial derivative. */
  def d(dx: Var): Vector
  /** Returns the Jacobian matrix. */
  def d(dx: VectorVar): Matrix

  def + (y: Vector): Vector = AddVector(this, y).simplify
  def - (y: Vector): Vector = AddVector(this, NegVector(y)).simplify
  def * (y: Scalar): Vector = ScalarVectorProduct(y, this).simplify
  def / (y: Scalar): Vector = ScalarVectorProduct(1/y, this).simplify

  def * (y: Vector): Scalar = InnerProduct(this, y).simplify
  def *~ (y: Vector): Matrix = OuterProduct(this, y).simplify

  def unary_+ : Vector = simplify
  def unary_- : Vector = NegVector(this).simplify
}

/** Vector of all 0's */
case class ZeroVector(size: IntScalar = IntConst("n")) extends Vector {
  override def toString: String = "0"
  override def d(dx: Var): Vector = this
  override def d(dx: VectorVar): Matrix = ZeroMatrix(size, dx.size)
  override def apply(env: Map[String, Tensor]): Vector = this
}

/** Vector of all 1's */
case class OneVector(size: IntScalar = IntConst("n")) extends Vector {
  override def toString: String = "1"
  override def d(dx: Var): Vector = ZeroVector(size)
  override def d(dx: VectorVar): Matrix = ZeroMatrix(size, dx.size)
  override def apply(env: Map[String, Tensor]): Vector = this
}

/** Vector value. */
case class VectorVal(x: Array[Double]) extends Vector {
  override def toString: String = x.mkString("[", ", ", "]")
  override def equals(that: Any): Boolean = that match {
    case VectorVal(y) => x.sameElements(y)
    case _ => false
  }
  override def size: IntScalar = IntVal(x.length)
  override def d(dx: Var): Vector = ZeroVector(size)
  override def d(dx: VectorVar): Matrix = ZeroMatrix(size, dx.size)
  override def apply(env: Map[String, Tensor]): VectorVal = this

  override def simplify: Vector = {
    if (x.forall(_ == 0)) ZeroVector(IntVal(x.length))
    else if (x.forall(_ == 1)) OneVector(IntVal(x.length))
    else this
  }
}

/** Constant vector. Different from VectorVal that has concrete values, this is of constant yet abstract value. */
case class ConstVector(symbol: String, size: IntScalar = IntConst("n")) extends Vector {
  override def toString: String = symbol
  override def d(dx: Var): Vector = ZeroVector(size)
  override def d(dx: VectorVar): Matrix = ZeroMatrix(size, dx.size)
  override def apply(env: Map[String, Tensor]): Vector = env.get(symbol) match {
    case Some(x: ZeroVector) => x
    case Some(x: OneVector) => x
    case Some(x: VectorVal) => x
    case x => throw new IllegalArgumentException(s"Invalid type: ${x.getClass}, expected 0 | 1 | VectorVal")
  }
}

/** Vector variable */
case class Vars(x: Scalar*) extends Vector {
  override def toString: String = x.mkString("[", ", ", "]")

  override def size: IntScalar = IntVal(x.length)

  override def d(dx: Var): Vector = {
    Vars(x.map(_.d(dx)): _*).simplify
  }

  override def d(dx: VectorVar): Matrix = {
    RowMatrix(x.map(_.d(dx).simplify): _*).simplify
  }

  override def apply(env: Map[String, Tensor]): Vector = {
    Vars(x.map(_(env)): _*).simplify
  }

  override def simplify: Vector = {
    if (x.forall(_.isInstanceOf[Val]))
      VectorVal(x.map(_.asInstanceOf[Val].x).toArray).simplify
    else
      this
  }
}

/** Abstract vector variable */
case class VectorVar(symbol: String, size: IntScalar = IntConst("n")) extends Vector {
  override def toString: String = symbol

  override def d(dx: Var): Vector = TangentVector(this, dx)

  override def d(dx: VectorVar): Matrix = {
    if (symbol.equals(dx.symbol)) IdentityMatrix(size, size) else ZeroMatrix(size, size)
  }

  override def apply(env: Map[String, Tensor]): Vector = env.get(symbol) match {
    case None => this
    case Some(x : Vector) => x
    case x => throw new IllegalArgumentException(s"Invalid type: ${x.getClass}, expected Vector")
  }
}

/** The derivative of a scalar y with respect to a vector x. */
case class GradientVector(y: Var, x: VectorVar) extends Vector {
  override def toString: String = s"\u2202${y.symbol}/\u2202${x.symbol}"

  override def size: IntScalar = x.size

  override def d(dx: Var): Vector = throw new UnsupportedOperationException("derivative of gradient vector")

  override def d(dx: VectorVar): Matrix = throw new UnsupportedOperationException("derivative of gradient vector")

  override def apply(env: Map[String, Tensor]): Vector = {
    val yv = env.get(y.symbol) match {
      case None | Some(Val(_)) => y
      case Some(y: Scalar) => y
      case a => throw new IllegalArgumentException(s"Invalid type: ${a.getClass}, expected Scalar")
    }

    env.get(x.symbol) match {
      case None => yv.d(x)
      case Some(x: VectorVar) => yv.d(x)
      case Some(Vars(x @ _*)) if x.forall(_.isInstanceOf[Var]) => yv.d(x.map(_.asInstanceOf[Var]): _*)
      case a => throw new IllegalArgumentException(s"Invalid type: ${a.getClass}, expected VectorVar or Vars")
    }
  }
}

/** The derivative of a vector y with respect to a scalar x. */
case class TangentVector(y: VectorVar, x: Var) extends Vector {
  override def toString: String = s"\u2202${y.symbol}/\u2202${x.symbol}"

  override def size: IntScalar = y.size

  override def d(dx: Var): Vector = throw new UnsupportedOperationException("derivative of tangent vector")

  override def d(dx: VectorVar): Matrix = throw new UnsupportedOperationException("derivative of tangent vector")

  override def apply(env: Map[String, Tensor]): Vector = {
    val yv = env.get(y.symbol) match {
      case None | Some(ZeroVector(_)) | Some(OneVector(_)) | Some(VectorVal(_)) | Some(ConstVector(_, _)) => y
      case Some(y: Vector) => y
      case a => throw new IllegalArgumentException(s"Invalid type: ${a.getClass}, expected Vector")
    }

    env.get(x.symbol) match {
      case None => yv.d(x)
      case Some(x: Var) => yv.d(x)
      case a => throw new IllegalArgumentException(s"Invalid type: ${a.getClass}, expected Var")
    }
  }
}

/** x + y */
case class AddVector(x: Vector, y: Vector) extends Vector {
  if (x.size != y.size) throw new IllegalArgumentException(s"Vector sizes mismatch: ${x.size} vs ${y.size}")

  override def toString: String = y match {
    case NegVector(y) => s"$x - $y"
    case _ => s"$x + $y"
  }

  override def equals(o: Any): Boolean = o match {
    case AddVector(a, b) => (x == a && y == b) || (x == b && y == a)
    case _ => false
  }

  override def size: IntScalar = x.size

  override def apply(env: Map[String, Tensor]): Vector = x(env) + y(env)

  override def d(dx: Var): Vector = {
    x.d(dx) + y.d(dx)
  }

  override def d(dx: VectorVar): Matrix = {
    x.d(dx) + y.d(dx)
  }

  override def simplify: Vector = (x, y) match {
    case (VectorVal(a), VectorVal(b)) => VectorVal(a.zip(b).map { case (x, y) => x + y })
    case (ZeroVector(_), b) => b
    case (a, ZeroVector(_)) => a
    case (a, NegVector(NegVector(b))) => a + b
    case (NegVector(a), NegVector(b)) => -(a + b)
    case (NegVector(a), b) if a == b => ZeroVector(a.size)
    case (a, NegVector(b)) if a == b => ZeroVector(a.size)
    case (ScalarVectorProduct(a, b), ScalarVectorProduct(c, d)) if b == d => (a + c) * b
    case (ScalarVectorProduct(a, b), ScalarVectorProduct(c, d)) if a == c => a * (b + d)
    case (ScalarVectorProduct(a, b), c) if b == c => (a + 1) * b
    case (a, ScalarVectorProduct(b, c)) if a == c => (b + 1) * a
    case (ScalarVectorProduct(a, b), NegVector(ScalarVectorProduct(c, d))) if b == d => (a - c) * b
    case (ScalarVectorProduct(a, b), NegVector(ScalarVectorProduct(c, d))) if a == c => a * (b - d)
    case (NegVector(ScalarVectorProduct(a, b)), c) if b == c => (1 - a) * b
    case (a, NegVector(ScalarVectorProduct(b, c))) if a == c => (1 - b) * a
    case (a, b) if a == b => 2 * a
    case _ => this
  }
}

/** -x */
case class NegVector(x: Vector) extends Vector {
  override def toString: String = x match {
    case AddVector(_, _) => s"-($x)"
    case _ => s"-$x"
  }

  override def size: IntScalar = x.size

  override def apply(env: Map[String, Tensor]): Vector = -x(env)

  override def d(dx: Var): Vector = {
    -x.d(dx)
  }

  override def d(dx: VectorVar): Matrix = {
    -x.d(dx)
  }

  override def simplify: Vector = x match {
    case a @ ZeroVector(_) => a
    case VectorVal(a) => VectorVal(a.map(-_))
    case NegVector(a) => a
    case _ => this
  }
}

/** a * x */
case class ScalarVectorProduct(a: Scalar, x: Vector) extends Vector {
  override def toString: String = {
    val xs = x match {
      case AddVector(_, _) => s"($x)"
      case _ => x.toString
    }

    val as = a match {
      case Add(_, _) | Sub(_, _) | Power(_, _) | InnerProduct(_, _) => s"($a)"
      case _ => a.toString
    }

    s"$as * $xs"
  }

  override def size: IntScalar = x.size

  override def apply(env: Map[String, Tensor]): Vector = a(env) * x(env)

  override def d(dx: Var): Vector = {
    a * x.d(dx) + a.d(dx) * x
  }

  override def d(dx: VectorVar): Matrix = {
    a * x.d(dx) + a.d(dx) *~ x
  }

  override def simplify: Vector = (a, x) match {
    case (Val(0), _) => ZeroVector(size)
    case (_, ZeroVector(_)) => ZeroVector(size)
    case (Val(1), b) => b
    case (Val(-1), b) => -b
    case (Val(a), VectorVal(b)) => VectorVal(b.map(_ * a))
    case (Neg(a), NegVector(b)) => a * b
    case (Neg(a), b) => -(a * b)
    case (a, NegVector(b)) => -(a * b)
    case (a, ScalarVectorProduct(b, c)) => (a * b) * c
    case _ => this
  }
}

/** Inner product (x * y) */
case class InnerProduct(x: Vector, y: Vector) extends Scalar {
  if (x.size != y.size) throw new IllegalArgumentException(s"Vector sizes mismatch: ${x.size} vs ${y.size}")

  override def toString: String = s"<$x, $y>"

  override def apply(env: Map[String, Tensor]): Scalar = x(env) * y(env)

  override def d(dx: Var): Scalar = {
    (x.d(dx) * y) + (x * y.d(dx))
  }

  override def d(dx: VectorVar): Vector = {
    (x.d(dx) * y) + (y.d(dx) * x)
  }

  override def simplify: Scalar = (x, y) match {
    case (ZeroVector(_), _) => Val(0)
    case (_, ZeroVector(_)) => Val(0)
    case (VectorVal(a), VectorVal(b)) => Val(a.zip(b).map { case (x, y) => x * y }.sum)
    case (ScalarVectorProduct(Val(a), b), c) => a * (b * c)
    case (a, ScalarVectorProduct(Val(b), c)) => b * (a * c)
    case (NegVector(a), NegVector(b)) => a * b
    case (a, NegVector(b)) => -(a * b)
    case (NegVector(a), b) => -(a * b)
    case _ => this
  }
}

/** Outer product (x ** y) */
case class OuterProduct(x: Vector, y: Vector) extends Matrix {
  if (x.size != y.size) throw new IllegalArgumentException(s"Vector sizes mismatch: ${x.size} vs ${y.size}")

  override def toString: String = {
    val xs = x match {
      case AddVector(_, _) | ScalarVectorProduct(_, _) => s"($x)"
      case _ => x.toString
    }

    val ys = y match {
      case AddVector(_, _) | ScalarVectorProduct(_, _) => s"($y)"
      case _ => y.toString
    }

    s"$xs \u2297 $ys"
  }

  override def size: (IntScalar, IntScalar) = (x.size, y.size)

  override def apply(env: Map[String, Tensor]): Matrix = x(env) *~ y(env)

  override def d(dx: Var): Matrix = {
    (x.d(dx) *~ y) + (x *~ y.d(dx))
  }

  override def simplify: Matrix = (x, y) match {
    case (ZeroVector(_), _) => ZeroMatrix(size)
    case (_, ZeroVector(_)) => ZeroMatrix(size)
    case (ScalarVectorProduct(Val(a), b), c) => a * (b *~ c)
    case (a, ScalarVectorProduct(Val(b), c)) => b * (a *~ c)
    case (NegVector(a), NegVector(b)) => a *~ b
    case (a, NegVector(b)) => -(a *~ b)
    case (NegVector(a), b) => -(a *~ b)
    case _ => this
  }
}

File: base/src/main/java/smile/tensor/Vector.java
/*
 * Copyright (c) 2010-2026 Haifeng Li. All rights reserved.
 *
 * SMILE is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SMILE is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SMILE. If not, see <https://www.gnu.org/licenses/>.
 */
package smile.tensor;

import java.lang.foreign.MemorySegment;
import java.util.Arrays;
import static smile.linalg.blas.cblas_h.*;

/**
 * Mathematical vector interface. Vectors are a specialized case of matrices.
 * Specifically, a vector is a matrix with either one row (a row-vector) or
 * one column (a column-vector). Column-vectors are the more common
 * representation and are often simply referred to as vectors.
 *
 * @author Haifeng Li
 */
public abstract class Vector extends DenseMatrix {
    /**
     * Default constructor for readObject.
     */
    Vector() {

    }

    /**
     * Constructor.
     * @param memory the memory segment of data.
     * @param m the number of rows.
     * @param n the number of columns.
     */
    Vector(MemorySegment memory, int m, int n) {
        super(memory, m, n, m, null, null);
        assert(m == 1 || n == 1);
    }

    /**
     * Returns the number of elements.
     * @return the number of elements.
     */
    public abstract int size();

    /**
     * Returns {@code A[i]}.
     * @param i the row index.
     * @return the cell value.
     */
    public abstract double get(int i);

    /**
     * Returns {@code A[i]} for Scala users.
     * @param i the index.
     * @return the cell value.
     */
    public double apply(int i) {
        return get(i);
    }

    /**
     * Sets {@code A[i] = x}.
     * @param i the index.
     * @param x the cell value.
     */
    public abstract void set(int i, double x);

    /**
     * Sets {@code A[i] = x} for Scala users.
     * @param i the index.
     * @param x the cell value.
     */
    public void update(int i, double x) {
        set(i, x);
    }

    /**
     * Sets {@code A[i] += x}.
     * @param i the index.
     * @param x the operand.
     */
    public abstract void add(int i, double x);

    /**
     * Sets {@code A[i] -= x}.
     * @param i the index.
     * @param x the operand.
     */
    public abstract void sub(int i, double x);

    /**
     * Sets {@code A[i] *= x}.
     * @param i the index.
     * @param x the operand.
     */
    public abstract void mul(int i, double x);

    /**
     * Sets {@code A[i] /= x}.
     * @param i the index.
     * @param x the operand.
     */
    public abstract void div(int i, double x);

    /**
     * Returns a slice of vector, which shares the data storage.
     * @param from the initial index of the range to be copied, inclusive.
     * @param to the final index of the range to be copied, exclusive.
     * @return a slice of this vector.
     */
    public abstract Vector slice(int from, int to);

    @Override
    public Vector column(int j) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Vector copy() {
        return copy(0, size());
    }

    /**
     * Copies the specified range of the vector into a new vector.
     * The final index of the range may be greater than vector length,
     * in which case 0 is placed in all extra elements.
     * @param from the initial index of the range to be copied, inclusive.
     * @param to the final index of the range to be copied, exclusive.
     *           This index may lie outside the array.
     * @return a copy of this vector.
     */
    public abstract Vector copy(int from, int to);

    /**
     * Copies the specified range of the vector into another vector.
     * @param src the destination vector.
     * @param pos starting position in this vector.
     * @param dest the destination vector.
     * @param destPos starting position in the destination vector.
     * @param length the number of vector elements to be copied.
     */
    public static void copy(Vector src, int pos, Vector dest, int destPos, int length) {
        if (src instanceof Vector32 a && dest instanceof Vector32 b) {
            System.arraycopy(a.array, pos, b.array, destPos, length);
        } else if (src instanceof Vector64 a && dest instanceof Vector64 b) {
            System.arraycopy(a.array, pos, b.array, destPos, length);
        } else {
            for (int i = 0; i < length; i++) {
                dest.set(destPos + i, src.get(pos + i));
            }
        }
    }

    /**
     * Returns an array containing all the elements in this vector.
     * @param a the array into which the elements of the vector are to be stored
     *          if it is big enough; otherwise, a new array is allocated.
     * @return an array containing the elements of the vector.
     */
    public abstract double[] toArray(double[] a);

    /**
     * Returns an array containing all the elements in this vector.
     * @param a the array into which the elements of the vector are to be stored,
     *          if it is big enough; otherwise, a new array is allocated.
     * @return an array containing the elements of the vector.
     */
    public abstract float[] toArray(float[] a);

    /**
     * Assigns the specified value to each element of the specified range of
     * the specified vector.
     * @param from the index of the first element (inclusive) to be filled
     *             with the specified value.
     * @param to the index of the last element (exclusive) to be filled
     *           with the specified value.
     * @param value the value to be stored in specified range of the vector.
     */
    public abstract void fill(int from, int to, double value);

    /**
     * Returns the matrix with the elements of this vector as the diagonal.
     * @return the diagonal matrix.
     */
    public DenseMatrix diagflat() {
        int n = size();
        DenseMatrix matrix = DenseMatrix.zeros(scalarType(), n, n);
        for (int i = 0; i < n; i++) {
            matrix.set(i, i, get(i));
        }
        return matrix;
    }

    /**
     * Returns a memory segment, at the given offset with the size specified
     * by the given argument.
     * @param array the primitive array backing the heap segment.
     * @param offset the base offset.
     * @param length the length.
     * @return the memory segment.
     */
    static MemorySegment memory(float[] array, int offset, int length) {
        var memory = MemorySegment.ofArray(array);
        if (offset != 0 || length != array.length) {
            long byteSize = 4;
            memory = memory.asSlice(offset * byteSize, length * byteSize);
        }
        return memory;
    }

    /**
     * Returns a memory segment, at the given offset with the size specified
     * by the given argument.
     * @param array the primitive array backing the heap segment.
     * @param offset the base offset.
     * @param length the length.
     * @return the memory segment.
     */
    static MemorySegment memory(double[] array, int offset, int length) {
        var memory = MemorySegment.ofArray(array);
        if (offset != 0 || length != array.length) {
            long byteSize = 8;
            memory = memory.asSlice(offset * byteSize, length * byteSize);
        }
        return memory;
    }

    /**
     * Creates a column vector.
     * @param array the primitive array backing the vector.
     * @return a column vector.
     */
    public static Vector column(double[] array) {
        return Vector64.column(array, 0, array.length);
    }

    /**
     * Creates a column vector.
     * @param array the primitive array backing the vector.
     * @return a column vector.
     */
    public static Vector column(float[] array) {
        return Vector32.column(array, 0, array.length);
    }

    /**
     * Creates a row vector.
     * @param array the primitive array backing the vector.
     * @return a row vector.
     */
    public static Vector row(double[] array) {
        return Vector64.row(array, 0, array.length);
    }

    /**
     * Creates a row vector.
     * @param array the primitive array backing the vector.
     * @return a row vector.
     */
    public static Vector row(float[] array) {
        return Vector32.row(array, 0, array.length);
    }

    /**
     * Returns a zero column vector.
     * @param scalarType the scalar type.
     * @param length the length of vector.
     * @return a zero column vector.
     */
    public static Vector zeros(ScalarType scalarType, int length) {
        return switch (scalarType) {
            case Float64 -> column(new double[length]);
            case Float32 -> column(new float[length]);
            default -> throw new UnsupportedOperationException("Unsupported ScalarType: " + scalarType);
        };
    }

    /**
     * Returns a zero column vector of the same scalar type as this vector.
     * @param length the length of vector.
     * @return a zero column vector.
     */
    public Vector zeros(int length) {
        return zeros(scalarType(), length);
    }

    /**
     * Returns a one column vector.
     * @param scalarType the scalar type.
     * @param length the length of vector.
     * @return a one column vector.
     */
    public static Vector ones(ScalarType scalarType, int length) {
        return switch (scalarType) {
            case Float64 -> {
                double[] array = new double[length];
                Arrays.fill(array, 1.0);
                yield column(array);
            }
            case Float32 -> {
                float[] array = new float[length];
                Arrays.fill(array, 1.0f);
                yield column(array);
            }
            default -> throw new UnsupportedOperationException("Unsupported ScalarType: " + scalarType);
        };
    }

    /**
     * Returns a one column vector of the same scalar type as this vector.
     * @param length the length of vector.
     * @return a one column vector.
     */
    public Vector ones(int length) {
        return ones(scalarType(), length);
    }

    /**
     * Returns the minimal elements of the vector.
     * @return the minimal elements of the vector.
     */
    public double min() {
        int length = size();
        double min = Double.MAX_VALUE;
        for (int i = 0; i < length; i++) {
            min = Math.min(min, get(i));
        }
        return min;
    }

    /**
     * Returns the maximal elements of the vector.
     * @return the maximal elements of the vector.
     */
    public double max() {
        int length = size();
        double max = Double.MIN_VALUE;
        for (int i = 0; i < length; i++) {
            max = Math.max(max, get(i));
        }
        return max;
    }

    /**
     * Sums the elements of the vector.
     * @return Sum of the elements of the vector.
     */
    public double sum() {
        int length = size();
        double s = 0;
        for (int i = 0; i < length; i++) {
            s += get(i);
        }
        return s;
    }

    /**
     * Returns the mean of the elements of the vector.
     * @return the mean of the elements of the vector.
     */
    public double mean() {
        int length = size();
        return length > 0 ? sum() / length : 0;
    }

    /**
     * The softmax function without overflow. The function normalizes
     * the vector into a probability distribution proportional to the
     * exponentials of the input numbers. That is, prior to applying
     * softmax, some vector components could be negative, or greater
     * than one; and might not sum to 1; but after applying softmax,
     * each component will be in the interval (0,1), and the components
     * will add up to 1, so that they can be interpreted as probabilities.
     * Furthermore, the larger input components will correspond to larger
     * probabilities.
     *
     * @return the index of largest posteriori probability.
     */
    public abstract int softmax();

    /**
     * Sums the absolute values of the elements of the vector.
     * @return Sum of the absolute values of the elements of the vector.
     */
    public double asum() {
        return switch(scalarType()) {
            case Float64 -> cblas_dasum(size(), memory, 1);
            case Float32 -> cblas_sasum(size(), memory, 1);
            default -> throw new UnsupportedOperationException("Unsupported scalar type: " + scalarType());
        };
    }

    /**
     * Computes a constant alpha times a vector x plus this vector y.
     * The result overwrites the initial values of this vector y.
     *
     * @param alpha If {@code alpha = 0} this routine returns without any computation.
     *
     * @param x Input vector.
     */
    public void axpy(double alpha, Vector x) {
        if (scalarType() != x.scalarType()) {
            throw new IllegalArgumentException("Different scalar type: " + x.scalarType());
        }
        if (size() != x.size()) {
            throw new IllegalArgumentException("Different vector size: " + x.size());
        }

        switch(scalarType()) {
            case Float64 -> cblas_daxpy(size(), alpha, x.memory, 1, memory, 1);
            case Float32 -> cblas_saxpy(size(), (float) alpha, x.memory, 1, memory, 1);
            default -> throw new UnsupportedOperationException("Unsupported scalar type: " + scalarType());
        }
    }

    /**
     * Computes the dot product of two vectors.
     *
     * @param x Input vector contains the second vector operand.
     *
     * @return dot product.
     */
    public double dot(Vector x) {
        if (scalarType() != x.scalarType()) {
            throw new IllegalArgumentException("Different scalar type: " + x.scalarType());
        }
        if (size() != x.size()) {
            throw new IllegalArgumentException("Different vector size: " + x.size());
        }

        return switch(scalarType()) {
            case Float64 -> cblas_ddot(size(), x.memory, 1, memory, 1);
            case Float32 -> cblas_sdot(size(), x.memory, 1, memory, 1);
            default -> throw new UnsupportedOperationException("Unsupported scalar type: " + scalarType());
        };
    }

    /**
     * Computes the Euclidean (L2) norm of a vector.
     *
     * @return Euclidean norm.
     */
    public double norm2() {
        return switch(scalarType()) {
            case Float64 -> cblas_dnrm2(size(), memory, 1);
            case Float32 -> cblas_snrm2(size(), memory, 1);
            default -> throw new UnsupportedOperationException("Unsupported scalar type: " + scalarType());
        };
    }

    /**
     * Computes the L1 norm of a vector.
     *
     * @return L1 norm.
     */
    public double norm1() {
        return switch(scalarType()) {
            case Float64 -> cblas_dasum(size(), memory, 1);
            case Float32 -> cblas_sasum(size(), memory, 1);
            default -> throw new UnsupportedOperationException("Unsupported scalar type: " + scalarType());
        };
    }

    /**
     * Computes the L-Infinity norm of a vector.
     *
     * @return L-Infinity norm.
     */
    public double normInf() {
        return Math.abs(get(iamax()));
    }

    @Override
    public Vector scale(double alpha) {
        switch(scalarType()) {
            case Float64 -> cblas_dscal(size(), alpha, memory, 1);
            case Float32 -> cblas_sscal(size(), (float) alpha, memory, 1);
            default -> throw new UnsupportedOperationException("Unsupported scalar type: " + scalarType());
        }
        return this;
    }

    /**
     * Swaps two vectors.
     *
     * @param x Vector to be swapped.
     */
    public void swap(Vector x) {
        if (scalarType() != x.scalarType()) {
            throw new IllegalArgumentException("Different scalar type: " + x.scalarType());
        }
        if (size() != x.size()) {
            throw new IllegalArgumentException("Different vector size: " + x.size());
        }

        switch(scalarType()) {
            case Float64 -> cblas_dswap(size(), x.memory, 1, memory, 1);
            case Float32 -> cblas_sswap(size(), x.memory, 1, memory, 1);
            default -> throw new UnsupportedOperationException("Unsupported scalar type: " + scalarType());
        }
    }

    /**
     * Searches a vector for the first occurrence of the maximum absolute
     * value.
     *
     * @return The first index of the maximum absolute value of vector x.
     */
    public int iamax() {
        return (int) switch(scalarType()) {
            case Float64 -> cblas_idamax(size(), memory, 1);
            case Float32 -> cblas_isamax(size(), memory, 1);
            default -> throw new UnsupportedOperationException("Unsupported scalar type: " + scalarType());
        };
    }
}
